src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs

```diff
@@ -997,50 +997,80 @@ private void RemoveWorseMembers<TMember>(ArrayBuilder<MemberResolutionResult<TMe
 
             var worse = ArrayBuilder<int>.GetInstance(results.Count, unknown);
 
-            for (int c1Idx = 0; c1Idx < results.Count; c1Idx++)
+            int c1Idx;
+            for (c1Idx = 0; c1Idx < results.Count; c1Idx++)
             {
-                var c1result = results[c1Idx];
+                if (results[c1Idx].Result.IsValid)
+                {
+                    worse[c1Idx] = betterThanEverything;
+                    break;
+                }
+            }
 
-                // If we already know this is worse than something else, no need to check again.
-                if (!c1result.Result.IsValid || worse[c1Idx] == worseThanSomething)
+            for (int c2Idx = c1Idx + 1; c2Idx < results.Count; c2Idx++)
+            {
+                if (!results[c2Idx].Result.IsValid)
                 {
                     continue;
                 }
 
-                bool c1IsBetterThanEverything = true;
-                for (int c2Idx = 0; c2Idx < results.Count; c2Idx++)
+                var better = BetterFunctionMember(results[c1Idx], results[c2Idx], arguments.Arguments, ref useSiteDiagnostics);
+                if (better == BetterResult.Left)
+                {
+                    worse[c2Idx] = worseThanSomething;
+                }
+                else if (better == BetterResult.Right)
+                {
+                    worse[c1Idx] = worseThanSomething;
+                    worse[c2Idx] = betterThanEverything;
+                    c1Idx = c2Idx;
+                }
+                else
                 {
-                    var c2result = results[c2Idx];
+                    worse[c1Idx] = notBetterThanEverything;
+                    worse[c2Idx] = notBetterThanEverything;
 
-                    if (!c2result.Result.IsValid)
+                    while(++c1Idx < results.Count)
                     {
-                        continue;
+                        if (results[c1Idx].Result.IsValid)
+                        {
+                            worse[c1Idx] = betterThanEverything;
```

Comment:
Never mind, I see the additional loop below that compares `results[c1Idx]` with all others.